### PR TITLE
feat: integrate RpcMiddleware in IPC

### LINKED ISSUE #7721: Integrate RpcMiddleware in IPC
### Describe the feature

currently the ipc server implementation does not include the new jsonrpsee RpcMiddleware

https://github.com/paradigmxyz/reth/blob/58cb524d73e1935a263ab8cd08777580cf67e15a/crates/rpc/ipc/src/server/mod.rs#L40-L40

which acts like a service over json rpc request objects

## TODO
1. integrate the trait into IpcServerBuilder, see also jsonrpsee::Server:
```
pub struct Server<HttpMiddleware = Identity, RpcMiddleware = Identity>
```

2. integrate the RpcMiddleware in ipc request handling (more difficult), see also `jsonrpsee::TowerServiceNoHttp`
https://github.com/paritytech/jsonrpsee/blob/5c65daf0b5c080b0c0cd984b3d522a5639f7d768/server/src/server.rs#L998

which is used like:

https://github.com/paritytech/jsonrpsee/blob/5c65daf0b5c080b0c0cd984b3d522a5639f7d768/server/src/server.rs#L1192-L1207

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by AbnerZheng**
close https://github.com/paradigmxyz/reth/issues/7721

### CODE REVIEW

File: crates/rpc/ipc/src/server/mod.rs [hunk 1]
```diff
@@ -21,31 +24,50 @@ use tokio::{
...
+    HttpMiddleware: Layer<TowerServiceNoHttp<RpcMiddleware>> + Send + 'static,
+    <HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service: Send
+        + Service<
+            String,
+            Response = Option<String>,
+            Error = Box<dyn std::error::Error + Send + Sync + 'static>,
+        >,
+    <<HttpMiddleware as Layer<TowerServiceNoHttp<RpcMiddleware>>>::Service as Service<String>>::Future:
+    Send + Unpin,
+{
```
Comment (mattsse) [inline crates/rpc/ipc/src/server/mod.rs:70]: we can relax this a bit and on require these bounds for functions that need it and don't require anything else for the impl block with functions that don't need it

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great!

this is the right direction!

File: crates/rpc/ipc/src/server/mod.rs (hunk 1 above)
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:70]: Would like to try it after clean up codes.

Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:77]: Is https://github.com/paradigmxyz/reth/pull/7790/commits/110e4c7f08f6430d3d96d9263ef0a99d7d6f676e  what you want?  

`IpcServer::start_inner` require these bounds, and `IpcServer::start` call ``IpcServer::start_inner` so both functions in the impl block need these bounds. So I move `IpcServer::endpoint` out.

File: crates/rpc/ipc/src/server/mod.rs [hunk 2]
```diff
@@ -244,16 +269,87 @@ pub(crate) struct ServiceData {
     pub(crate) method_sink: MethodSink,
 }
 
+/// Similar to [`tower::ServiceBuilder`] but doesn't
+/// support any tower middleware implementations.
+#[derive(Debug, Clone)]
+pub struct RpcServiceBuilder<L>(tower::ServiceBuilder<L>);
```
Comment (mattsse) [inline crates/rpc/ipc/src/server/mod.rs:275]: do we need this type?
looks like this is the same as the jsonrpsee type, so we can use that instead?

**Review by mattsse — CHANGES_REQUESTED** [review]
this is great!
only have a few questions

we also want to configure metrics again:

https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-builder/src/lib.rs#L1759-L1765

File: crates/rpc/ipc/src/server/rpc_service.rs [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        middleware::rpc::{ResponseFuture, RpcServiceT},
+        IdProvider,
+    },
+    types::{error::reject_too_many_subscriptions, ErrorCode, ErrorObject, Request},
+    BoundedSubscriptions, ConnectionDetails, MethodCallback, MethodResponse, MethodSink, Methods,
+    SubscriptionState,
+};
+use std::sync::Arc;
+
+/// JSON-RPC service middleware.
```
Comment (mattsse) [inline crates/rpc/ipc/src/server/rpc_service.rs:14]: this is the same as the jsonrpsee type?

can we reuse that instead?

Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/rpc_service.rs:14]: Yes, it is totally the same as the [`jsonrpsee::RpcService`](https://github.com/paritytech/jsonrpsee/blob/d3b1b3afa8b94d9f6f938221358d697286a81b9a/server/src/middleware/rpc/layer/rpc_service.rs#L43), but its field is private and the `RpcService::new` function is `pub(crate)`, furthermore the type of its field `cfg` is `RpcServiceCfg` which is `pub(crate)`, so I am afraid I can't reuse it, right?

File: crates/rpc/ipc/src/server/mod.rs (hunk 2 above)
Comment (AbnerZheng) [reply crates/rpc/ipc/src/server/mod.rs:275]: Yes, they are the same, but `RpcServiceBuilder::service` is `pub(crate)`, and I have no other way to setting inner service other than this method. 
https://github.com/paritytech/jsonrpsee/blob/98675a06ee0d386256954c4fa21b3fb02345d045/server/src/middleware/rpc/mod.rs#L102

**Comment by AbnerZheng** [comment]
> we also want to configure metrics again:
> 
> https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc-builder/src/lib.rs#L1759-L1765

Done.

**Review by mattsse — APPROVED** [review]
nice work

perhaps we can make some changes upstream but for now, duplicating the types is reasonable

**Pull request merged**